sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1602,12 +1595,6 @@ object EliminateSorts extends Rule[LogicalPlan] {
     plan match {
       case Sort(_, global, child) if canRemoveGlobalSort || !global =>
         recursiveRemoveSort(child, canRemoveGlobalSort)
-      case Sort(sortOrder, true, child) =>
-        // For this case, the upper sort is local so the ordering of present sort is unnecessary,
-        // so here we only preserve its output partitioning using `RepartitionByExpression`.
-        // We should use `None` as the optNumPartitions so AQE can coalesce shuffle partitions.
-        // This behavior is same with original global sort.
-        RepartitionByExpression(sortOrder, recursiveRemoveSort(child, true), None)
```

Comment:
Hmm, previously this rule looks into this global Sort's child to remove local and global Sort recursively without condition. But in the new `RemoveRedundantSorts` rule:

```scala
case s @ Sort(orders, true, child) =>
  val newChild = recursiveRemoveSort(child, optimizeGlobalSort = false)
```
`recursiveRemoveSort` in `RemoveRedundantSorts` only removes local Sort if its child is already sorted. Do we miss this optimization?

Reply:
`EliminateSorts` still does this job: https://github.com/apache/spark/pull/44429/files#diff-11264d807efa58054cca2d220aae8fba644ee0f0f2a4722c46d52828394846efR1577

Reply:
Said there are Sorts like

```
- Sort (local)
  - Sort (global)
    - Sort (local)
```

We reach:

```scala
case s @ Sort(_, global, child) => s.copy(child = recursiveRemoveSort(child, global))
```

Previously we can get rid of the middle global Sort and the bottom local Sort by `RepartitionByExpression(sortOrder, recursiveRemoveSort(child, true), None)` and:

```scala
case Sort(_, global, child) if canRemoveGlobalSort || !global =>
  recursiveRemoveSort(child, canRemoveGlobalSort)
```

How does `EliminateSorts` still do it?
The code you point is same (not changed in this PR):

```scala
case s @ Sort(_, global, child) => s.copy(child = recursiveRemoveSort(child, global))
```

But in `recursiveRemoveSort`, as `canRemoveGlobalSort` is false, we don't get rid of the middle global Sort now (it will be done in `RemoveRedundantSorts` now).

Do I miss ore misread something?

Reply:
After running `EliminateSorts`, the bottom sort is removed, then we `RemoveRedundantSorts` which will turn the middle sort to local sort.